### ISSUE TITLE
CLR transform doesn't include relative abundance step

### ISSUE THREAD

**Issue by exod1a** [microbiome/mia#91]
Hi
I noticed that when computing the CLR transform from the transformCounts() function and comparing it to the result from when I first calculate the relative abundances and then apply the CLR transformation, the results were different. Leo said it would be good to mention this to see if you want to keep it this way or change it. Here is an example that reproduces what I mentioned above.

```
library(mia)

data("GlobalPatterns")
tse <- GlobalPatterns

tse1 <- relAbundanceCounts(tse)
tse1 <- transformCounts(tse1, abund_values = "relabundance", method = "clr", pseudocount = 1)

tse2 <- transformCounts(tse, abund_values = "counts", method = "clr", pseudocount = 1)

assays(tse1)$clr == assays(tse2)$clr
```

**Comment by FelixErnst** [comment]
@tvborm

Can you have a look at it?

**Comment by antagomir** [comment]
To me it seems that the default behavior should be so that "compositionality + clr" gives the same result than "clr" when applied to count data. If there are other views, I would be curious to hear the justifications - @tvborm can check we had just missed to include compositionality transformation as a preprocessing step as part pf the clr procedure (should be I think).

**Comment by FelixErnst** [comment]
Just to rule out a tolerance issue: Can you post the upper left corner of each assay? 

Please also include the same output for `assay(tse1,"clr") - assay(tse2, "clr")`

**Comment by FelixErnst** [comment]
This is due to pseudocount being applied.

**Comment by antagomir** [comment]
I thought so. I think that this should give a warning if user is trying to add pseudocount 1 on relative abudances?

**Comment by FelixErnst** [comment]
Should be fixed now.

**Comment by FelixErnst** [comment]
@tvborm The snakemake run took longer than expected. So I had the time. I hope you don't mind.

**Comment by TuomasBorman** [comment]
@FelixErnst that's totally fine, thanks for fixing it

**Comment by asierFernandezP** [comment]
Hi,

I am also having some problems with CLR transformation. When doing CLR transformation on relative abundance values (Metaphan results removing the unknown portion and recalculating these relative abundances so that each sample has a total abundance = 1) and adding as peudocounts half of the minimum value, I get very different results on downstream analyses (including for instance the PERMANOVA R2, or species coefficients, which are quite high (between -4 and 4)) compared to this same analysis with Bray-Curtis distance (without CLR transformation). 

On the contrary, when doing CLR transformation on original relative abundance values from metaphlan results after removing the unknown portion (here I did not recalculate the relative abundances, so each sample has a total abundance of 100 - fraction corresponding to unknown) and adding 1 as pseudocount (I was doing this in the beginning until I realised it was wrong) I get very similar results to the analysis with Bray-Curtis. 

I am sorry I am not providing any more information, but does this make any sense? I

Thank you!

**Comment by TuomasBorman** [comment]
Hi!

Thanks for pointing this out! That is because `transformSamples( method = "clr", ...)` does relative transformation inside the function. In your case, relative transformation is performed to values that are already relative... 

@antagomir Any ideas how to fix this?

1. We cannot rely that all colSums equal to 1 (to check if values ae relative)
2. We cannot rely that the name of an assay is "relabundance"
3. We cannot rely that if all values are between 0-1, they are relative

However, we could add there a hidden option "`relative = FALSE`". We could check that if all values are between 0-1 and give a message. "It looks like values arerelative. if so, please ue relative = TRUE"....

-Tuomas

**Comment by asierFernandezP** [comment]
I am specifically using `transformCounts` function. Does it also do a relative transformation?

**Comment by TuomasBorman** [comment]
Yep, transformCounts does exactly the same thing as transfromSamples

**Comment by asierFernandezP** [comment]
Thank you! I think adding that  hidden option "relative = FALSE" could be really useful.

**Comment by TuomasBorman** [comment]
Hmm.. I have to still check it in more detail

There are 2 things that lead to different results (relative abundances & CLR vs counts & CLR)

@antagomir : 

> default behavior should be so that "compositionality + clr" gives the same result than "clr" 

--> since pseudocount is applied before relative transformation and because we don't know library sizes, ` min( assay(tse, "relabundance")[ assay(tse, "relabundance") > 0 ])` is the closest approximation compared to when CLR is applied to counts and pseudocount is 1


Another thing is that metaphlan's relative abudances do not necessarily sum up to 1. (Otherwise relative transf. to relative transf. == relative tranfs. [that is probably why it was done for all])

--> `relative ==option` is justified



For counts it goes like this

1. add pseudocount
2. apply relative transformation
3. do CLR

For relative abundances
1. add pseudocount 
(2. apply relative transformation [this will be optional for now on] )
2. Do CLR

**Comment by antagomir** [comment]
Important observation.

If I gather it correctly from here, this concerns cases where the colSums do not sum to 1 because some taxa have been left out? In such case, one may nevertheless, in some cases, want to rescale the data to relative abundance where colSums are 1. This is sometimes done after filtering out taxa. Although I think the recommended practice would be to have a category "Other" for the leave-out taxa, so that one can keep that information on relative abundances. Hence, even with relative abundances (like "relative=TRUE") we do not automatically want to add some solution. Instead, the user should explicitly choose whether to add relative transformation.  One option i to require that the data is readily provided in a suitable format. If the user wants to use relative data, they need to do transformation with their chosen methods _before_ CLR. My suggestion is therefore:

0) By default we assume that users already have the data in a suitable format for CLR, and minimal default processing is added
1) If input data has only positive values (no zeroes) then warning is generated if user tries to add pseudocount. There should be no need for pseudocount in general in positive data.
2) If input data is integers (counts), then relative transformation is done as a necessary preprocessing step before running clr. We cannot calculate clr for count data without such step. Or: we can also require that CLR is only run with data that is in [0,1] and generates error with integer data. This way we can make sure that the users always need to think about it and provide correct type of data. Examples can show how to do CLR in two steps (count -> relative -> clr). 
3) If input data is between [0, 1] then warning is generated if the user is trying to use pseudocount
4) If input data is between [0, 1] but does not sum to 1, then warning is generated if colSums do not equal to 1 but no relative transformation is done (user can add this if they need to)

Downside of these rules is that it makes handling a bit complex. On the other hand this should correspond to typical use cases and catch errors that might be done by inexperienced users.

Or do you have other suggestions?

**Comment by TuomasBorman** [comment]
Hmmm, yes... 

Well I haven't used metaphlan data that much, but if I remember, the example data that I used to test the importer did not have all the taxa (colSums != 1)

In my opinion, we should not modify too much the functionality with abund_values = "counts" as it works well. Currently the function should also work for relative abundances when correct pseudocount is added

I think those warnings are good. The real problem is when (colSums != 1) and that is why we should also have option for disabling relative transformation

**Comment by antagomir** [comment]
Hmmm. Thinking about it now, the definition of CLR does not demand relative transformation. It does assume simplex, where the values sum up to a fixed constant, which is often defined as 1 for convenience. But the definition of the clr transformation does not perform relative transformation, hence I think we should simply remove the relative transformation line from the .calc_clr function (and I should also fix the current contribution to vegan).

This being said, the relative transformation should not have any effect on the clr transformation. Relative transformation is just scaling of the values, and clr is invariant to the scale. I would remove the relative transformation step entirely but at the same this should not influence the outcome. 

Can we have a code example of a case where the outcomes are different, where this is not expected? Is this a problem with pseudocount, rather than relative transformation?

**Comment by TuomasBorman** [comment]
#' Gloor GB, Macklaim JM, Pawlowsky-Glahn V & Egozcue JJ (2017)
#' Microbiome Datasets Are Compositional: And This Is Not Optional.
#' Frontiers in Microbiology 8: 2224. doi: 10.3389/fmicb.2017.02224
https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full

1. Yes, we should remove relative transformation from .calc_clr
2. Although we should still check that 

- there are only positive values --> error if not
- colSums are equal --> warning if not

**Comment by antagomir** [comment]
Yes! colSums check and warning is also appropriate for count data, so the user will notice and make an informed choice.

**Comment by antagomir** [comment]
Also check the same for `rclr`.

**Comment by antagomir** [comment]
The _vegan_ implementation was OK already.

Let us also add comprehensive unit tests in _mia_ that cover the cases where the data is 

1. counts
2. values in [0,1]
3. with colSums = 1
4. with colSums = k (something else than 1..)
5. with colSums != 1 (variable colSums)

**Comment by antagomir** [comment]
Fixed now in #262 @asierFernandezP your testing & feedback is welcome..!

Development version installs with `devtools::install_github("microbiome/mia")`

**Comment by asierFernandezP** [comment]
Thank you for fixing it so quickly!!
